prusti-server/src/process_verification.rs

```diff
@@ -120,11 +266,129 @@ pub fn process_verification_request<'v, 't: 'v>(
+    viper_arc: &Arc<Viper>,
+    sender: mpsc::Sender<ServerMessage>,
+) -> VerificationResultKind {
+    let mut result_type = VerificationResultKind::Success;
+
+    // get the reporter global reference outside of the thread scope because it needs to
+    // be dropped by thread attached to the jvm. This is also why we pass it as reference
+    // and not per value
+    let env = &verification_context.env();
+    let jni = JniUtils::new(env);
+    let verifier_wrapper = silver::verifier::Verifier::with(env);
+    let reporter = jni.unwrap_result(verifier_wrapper.call_reporter(*verifier.verifier_instance()));
+    let rep_glob_ref = env.new_global_ref(reporter).unwrap();
+
+    debug!("Starting viper message polling thread");
+
+    // start thread for polling messages
+    thread::scope(|scope| {
+        let polling_thread = scope.spawn(|| polling_function(viper_arc, &rep_glob_ref, sender));
+        result_type = verifier.verify(viper_program);
+        polling_thread.join().unwrap();
+    });
+    debug!("Viper message polling thread terminated");
+    result_type
+}
+
+fn polling_function(
+    viper_arc: &Arc<Viper>,
+    rep_glob_ref: &jni::objects::GlobalRef,
+    sender: mpsc::Sender<ServerMessage>,
+) {
+    let verification_context = viper_arc.attach_current_thread();
+    let env = verification_context.env();
+    let jni = JniUtils::new(env);
+    let reporter_instance = rep_glob_ref.as_obj();
+    let reporter_wrapper = silver::reporter::PollingReporter::with(env);
+    loop {
+        while reporter_wrapper
+            .call_hasNewMessage(reporter_instance)
+            .unwrap()
+        {
+            let msg = reporter_wrapper
+                .call_getNewMessage(reporter_instance)
+                .unwrap();
+            debug!("Polling thread received {}", jni.class_name(msg).as_str());
+            match jni.class_name(msg).as_str() {
+                "viper.silver.reporter.QuantifierInstantiationsMessage" => {
+                    let msg_wrapper = silver::reporter::QuantifierInstantiationsMessage::with(env);
+                    let q_name =
+                        jni.get_string(jni.unwrap_result(msg_wrapper.call_quantifier(msg)));
+                    let q_inst = jni.unwrap_result(msg_wrapper.call_instantiations(msg));
+                    debug!("QuantifierInstantiationsMessage: {} {}", q_name, q_inst);
+                    // also matches the "-aux" and "_precondition" quantifiers generated
+                    // we encoded the position id in the line and column number since this is not used by
+                    // prusti either way
+                    if q_name.starts_with("prog.l") {
+                        let no_pref = q_name.strip_prefix("prog.l").unwrap();
+                        let stripped = no_pref
+                            .strip_suffix("-aux")
+                            .or(no_pref.strip_suffix("_precondition"))
+                            .unwrap_or(no_pref);
```

Comment:
Where do these alternatives come from?

Reply:
I did not investigate too much, but as far as I remember they are introduced by Z3